### PR TITLE
Minimal exception handling extension

### PR DESCRIPTION
**Pull request by dhil**
This patch extends the interpreter with a minimal exception handling facility. It is minimal in the sense that it supports only a single exception which can be thrown and caught. I think this should be sufficient to provide a base for further experiments.
    
The patch is implemented on top of the latest reference-types proposal, though in retrospect it seems to be unnecessary as with this minimal design there is no notion of exception types, and hence no need for reference types.

### CODE REVIEW

**Comment by dhil** [comment]
I can potentially disentangle the exception handling extension from the reference-types commits.

**Comment by rossberg** [comment]
Yes, I'd suggest separating the upstream merge from the exception PR. I just created #2, so maybe it's sufficient if you approve that and then this PR automatically adopts that as baseline -- though there might be some merge conflicts.

**Comment by dhil** [comment]
Sounds good to me! I will give it a try.

**Comment by dhil** [comment]
@rossberg I have updated the patch now. It is now a single commit on top of master with typed-func-refs merged in.

File: interpreter/syntax/free.ml [hunk 1]
```diff
@@ -123,6 +123,8 @@ let rec instr (e : instr) =
     memories zero
   | MemoryInit x -> memories zero ++ datas (idx x)
   | DataDrop x -> datas (idx x)
+  | Try (_, es1, es2) -> block es1 ++ block es2
```
Comment (rossberg) [inline interpreter/syntax/free.ml]:
```suggestion
  | Try (bt, es1, es2) -> block_type bt ++ block es1 ++ block es2
```

File: interpreter/text/arrange.ml [hunk 2]
```diff
@@ -288,6 +288,10 @@ let rec instr e =
     | Unary op -> unop op, []
     | Binary op -> binop op, []
     | Convert op -> cvtop op, []
+    | Try (bt, es1, es2) ->
+      "try", block_type bt @
+               [Node ("then", list instr es1); Node ("catch", list instr es2)]
```
Comment (rossberg) [inline interpreter/text/arrange.ml]:
```suggestion
        [Node ("do", list instr es1); Node ("catch", list instr es2)]
```

File: interpreter/text/parser.mly [hunk 3]
```diff
@@ -692,6 +698,43 @@ call_expr_results :
...
+      FuncType (ins' @ ins, out), es }
+try_block_result_body :
+  | try_ { fun c -> FuncType ([], []), $1 c }
+  | LPAR RESULT value_type_list RPAR try_block_result_body
+    { fun c ->
+      let FuncType (ins, out), es = $5 c in
+      let out' = snd $3 c in
+      FuncType (ins, out' @ out), es }
+try_ :
+ | LPAR THEN instr_list RPAR LPAR CATCH instr_list RPAR
```
Comment (rossberg) [inline interpreter/text/parser.mly]: Current proposal uses `do` for this. (May need a new token)
```suggestion
 | LPAR DO instr_list RPAR LPAR CATCH instr_list RPAR
```

File: interpreter/binary/encode.ml [hunk 4]
```diff
@@ -428,6 +428,8 @@ let encode m =
       | Convert (F64 F64Op.DemoteF64) -> assert false
       | Convert (F64 F64Op.ReinterpretInt) -> op 0xbf
 
+      | Try _ | Throw -> failwith "encode.ml: not yet implemented"
```
Comment (rossberg) [inline interpreter/binary/encode.ml]:
```suggestion
      | Try _ | Throw -> failwith "encode: not yet implemented"
```

File: interpreter/exec/eval.ml [hunk 5]
```diff
@@ -525,7 +528,17 @@ let rec step (c : config) : config =
...
       | Convert cvtop, Num n :: vs' ->
         (try Num (Eval_numeric.eval_cvtop cvtop n) :: vs', []
-        with exn -> vs', [Trapping (numeric_error e.at exn) @@ e.at])
+         with exn -> vs', [Trapping (numeric_error e.at exn) @@ e.at])
+
+      | Try (bt, es1, es2), vs ->
+        let FuncType (ts1, ts2) = block_type c.frame.inst bt e.at in
+        let n1 = List.length ts1 in
+        let n2 = List.length ts2 in
+        let args, vs' = take n1 vs e.at, drop n1 vs e.at in
```
Comment (rossberg) [inline interpreter/exec/eval.ml]:
```suggestion
        let args, vs' = split n1 vs e.at in
```

File: interpreter/exec/eval.ml [hunk 6]
```diff
@@ -616,6 +635,20 @@ let rec step (c : config) : config =
...
         args @ args' @ vs', [Invoke f' @@ e.at]
       )
+
+    | Throwing, _ ->
+       Uncaught.error e.at "uncaught exception"
+
+    | Catch (_, _, (_, ({it = Trapping _ | Breaking _ | Returning _; _} as e) :: _)), vs ->
+      vs, [e]
+
+    | Catch (n, es1, (_, { it = Throwing; at} :: _)), vs ->
```
Comment (rossberg) [inline interpreter/exec/eval.ml]:
```suggestion
    | Catch (n, es1, (_, {it = Throwing; _} :: _)), vs ->
```

File: interpreter/exec/eval.ml [hunk 7]
```diff
@@ -616,6 +635,20 @@ let rec step (c : config) : config =
...
+      let exn = [] in
+      vs, [Label (n, [], (exn, List.map plain es1)) @@ e.at]
```
Comment (rossberg) [inline interpreter/exec/eval.ml]:
```suggestion
      vs, [Label (n, [], ([], List.map plain es1)) @@ e.at]
```

**Comment by dhil** [comment]
Thanks!

**Comment by dhil** [comment]
@rossberg I have added a few test cases now (and a little bit of infrastructure to support testing, e.g. `assert_uncaught`).

**Comment by dhil** [comment]
Compiles and passes the test suite now. 

To make it pass the test suite I had to implement encode/decode for exceptions and their handlers. I have used the opcodes from the current iteration of the exception handling proposal.

File: interpreter/script/js.ml [hunk 8]
```diff
@@ -529,6 +529,8 @@ let of_assertion mods ass =
     of_assertion' mods act "assert_trap" [] None
   | AssertExhaustion (act, _) ->
     of_assertion' mods act "assert_exhaustion" [] None
+  | AssertUncaught (act, _) ->
+    of_assertion' mods act "assert_uncaught" [] None
```
Comment (rossberg) [inline interpreter/script/js.ml:540]: You'll need to add that function to the JS harness above.

File: interpreter/script/run.ml [hunk 9]
```diff
@@ -446,6 +447,13 @@ let run_assertion ass =
     | _ -> Assert.error ass.at "expected exhaustion error"
     )
 
+  | AssertUncaught (act, re) ->
+    trace ("Asserting trap...");
```
Comment (rossberg) [inline interpreter/script/run.ml]:
```suggestion
    trace ("Asserting exception...");
```

File: interpreter/script/run.ml [hunk 10]
```diff
@@ -446,6 +447,13 @@ let run_assertion ass =
...
+    (match run_action act with
+    | exception Eval.Uncaught (_, msg) -> assert_message ass.at "runtime" msg re
+    | _ -> Assert.error ass.at "expected runtime error"
```
Comment (rossberg) [inline interpreter/script/run.ml]:
```suggestion
    | _ -> Assert.error ass.at "expected exception"
```

File: interpreter/text/lexer.mll [hunk 11]
```diff
@@ -396,6 +396,7 @@ rule token = parse
   | "catch" { CATCH }
   | "do" { DO }
   | "throw" { THROW }
+  | "assert_uncaught" { ASSERT_UNCAUGHT }
```
Comment (rossberg) [inline interpreter/text/lexer.mll]: Nit: group with other assertions.

File: interpreter/text/parser.mly [hunk 12]
```diff
@@ -228,7 +228,7 @@ let inline_func_type_explicit (c : context) x ft at =
 %token NAN
 %token INPUT OUTPUT
 %token EOF
-%token TRY CATCH DO THROW
+%token TRY CATCH DO THROW ASSERT_UNCAUGHT
```
Comment (rossberg) [inline interpreter/text/parser.mly]: Nit: group with other assertions

File: interpreter/binary/decode.ml [hunk 13]
```diff
@@ -270,7 +270,17 @@ let rec instr s =
     end
 
   | 0x05 -> error s pos "misplaced ELSE opcode"
-  | 0x06| 0x07 | 0x08 | 0x09 | 0x0a as b -> illegal s pos b
+  | 0x06 ->
+     let bt = block_type s in
```
Comment (rossberg) [inline interpreter/binary/decode.ml]: Nit: indent -= 1

**Review by rossberg — COMMENTED** [review]
Nice! A couple more comments about that.

**Comment by dhil** [comment]
@rossberg I had forgotten to add a reduction rule for the success-case when the body of `try` raises no exceptions. My latest commit adds that rule. With this, I think the patch is complete.

File: interpreter/script/js.ml [hunk 14]
```diff
@@ -189,6 +189,13 @@ function assert_return(action, ...expected) {
     }
   }
 }
+
+function assert_uncaught(action) {
+  try { action() } catch (e) {
+    if (e instanceof WebAssembly.RuntimeError) return;
```
Comment (rossberg) [inline interpreter/script/js.ml]:
```suggestion
    if (!(e instanceof WebAssembly.RuntimeError)) return;
```

**Review by rossberg — APPROVED** [review]

**Pull request merged**